atspi-common/src/object_match.rs

```diff
@@ -0,0 +1,281 @@
+		self.ifaces = interfaces.iter().copied().collect::<InterfaceSet>();
+		self.ifaces_mt = mt;
+		self
+	}
+
+	/// Sets the inversion of the `MatchRule`, defaults to `false`, no inversion.
+	#[must_use]
+	pub fn invert(mut self, invert: bool) -> Self {
+		self.invert = invert;
+		self
+	}
+
+	/// Builds the [`ObjectMatchRule`]
+	///
+	/// [`ObjectMatchRule``]: crate::object_match::ObjectMatchRule
+	#[must_use]
+	pub fn build(self) -> ObjectMatchRule {
+		ObjectMatchRule {
+			states: self.states,
+			states_mt: self.states_mt,
+			attr: self.attr,
+			attr_mt: self.attr_mt,
+			roles: self.roles,
+			roles_mt: self.roles_mt,
+			ifaces: self.ifaces,
+			ifaces_mt: self.ifaces_mt,
+			invert: self.invert,
+			phantom: PhantomData,
+		}
+	}
+}
+
+/// Enumeration used by [`ObjectMatchRule`] to specify how to interpret [`ObjectRef`] objects.
+///
+/// [`ObjectRef`]: crate::object_ref::ObjectRef
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize, Type, Default)]
+#[repr(i32)]
+pub enum MatchType {
+	#[default]
+	/// Invalidates match criterion.
+	Invalid,
+
+	/// All of the criteria must be met.
+	All,
+
+	/// Any of the criteria must criteria must be met.
+	Any,
+
+	/// None of the criteria must be met.
+	NA,
+
+	/// Same as [`Self::All`] if the criterion item is non-empty - All of the criteria must be met.
+	/// For empty criteria this rule requires the returned value to also have empty set.
+	Empty,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize, Type)]
+#[repr(u32)]
+/// Enumeration used by interface `CollectionProxy` to specify the way [`ObjectRef`]
+/// objects should be sorted.
+///
+/// # Note
+///
+/// Only `Canonical` and `ReverseCanonical` are currently implemented.
+///
+/// see: [issue 140. "Are all the AtspiCollectionSortOrder types really a thing?"](https://gitlab.gnome.org/GNOME/at-spi2-core/-/issues/140)
+///
+/// [`ObjectRef`]: crate::object_ref::ObjectRef
+pub enum SortOrder {
+	/// Invalid sort order
+	///
+	/// Unimplemented
+	Invalid,
+
+	/// Canonical sort order
+	Canonical,
+
+	/// Flow sort order
+	///
+	/// Unimplemented
+	Flow,
+
+	/// Tab sort order
+	///
+	/// Unimplemented
+	Tab,
+
+	/// Reverse canonical sort order
+	ReverseCanonical,
+
+	/// Reverse flow sort order
+	///
+	/// Unimplemented
+	ReverseFlow,
+
+	/// Reverse tab sort order
+	///
+	/// Unimplemented
+	ReverseTab,
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use crate::{SortOrder, State};
+	use zbus_lockstep::method_args_signature;
+
+	#[test]
+	fn validate_match_rule_signature() {
+		let signature = method_args_signature!(member: "GetMatchesTo", interface: "org.a11y.atspi.Collection", argument: "rule");
+		assert_eq!(ObjectMatchRule::signature(), signature);
+	}
+
+	#[test]
+	fn validate_match_type_signature() {
+		let rule_signature = method_args_signature!(member: "GetMatchesTo", interface: "org.a11y.atspi.Collection", argument: "rule");
+		let match_type_signature = rule_signature.slice(3..4);
```

Comment:
Just add a comment here explaining why you're comparing these random indecies—I know why now, but future me is not as knowledgeable about the internals of `Collection`.